issue119

Taktar opened the pull request:
Closes #119.

### Summary
Adds a small backend endpoint that serves a merged JSON for a boneset, and checks in the first merged file for **Bony Pelvis**. This is used by the frontend and future work to fetch a complete payload from one place.

---

## 1) What was Changed
- **New file**: `boneset-api/data/final_bony_pelvis.json` (merged data for Bony Pelvis).
- **server.js**:
  - `const fs = require("fs/promises");`
  - `const DATA_DIR = path.join(__dirname, "data");`
  - **New route** `GET /api/boneset/:bonesetId` → reads `data/final_<bonesetId>.json` and returns it.
  - 404 handling if the file doesn’t exist.
- Existing routes remain unchanged:
  - `GET /combined-data`
  - `GET /api/description/`

## 2) Why it was Changed
The frontend needs a reliable, single source of truth for a full boneset payload (instead of assembling multiple files client-side). This endpoint meets the acceptance criteria in #119 and will be extended with more merged files in future sprints.

## 3) How it was Changed (technical)
- Reads local JSON from `boneset-api/data/` using `fs/promises`.
- Returns the file as `application/json` with no transformation.
- If a requested boneset file is missing, responds with `404 { "error": "Boneset '<id>' not found" }`.
- No breaking changes to existing endpoints or the frontend; this is additive.

## 4) Screenshots / Evidence
Manual verification (all successful):

```bash
# health
curl http://127.0.0.1:8000/
# -> {"message":"Welcome to the Boneset API (GitHub-Integrated)"}

# existing combined-data still works
curl http://127.0.0.1:8000/combined-data

# new endpoint (acceptance criteria)
curl http://127.0.0.1:8000/api/boneset/bony_pelvis
# -> returns full content of boneset-api/data/final_bony_pelvis.json

# error case
curl -i http://127.0.0.1:8000/api/boneset/not_real
# -> HTTP/1.1 404 with {"error":"Boneset 'not_real' not found"}

Taktar commented:
> Great Job so far in the work you have done, review the comments i made and implement the changes. Testing is a very important step in making sure that what we have done works as expected so while we don't have an automated backend testing suite set up yet, it's important to manually verify the functionality.
> 
> Please add a comment to this pull request detailing the steps you took to test the new endpoint. For example:
> 
> Confirmed that navigating to http://127.0.0.1:8000/api/boneset/bony_pelvis in the browser successfully displays the JSON content.
> 
> Confirmed that navigating to a non-existent boneset (e.g., /api/boneset/invalid_id) returns the expected 404 Not Found error.

Manual Test Plan

1) Success path
curl -sS http://127.0.0.1:8000/api/boneset/bony_pelvis | jq -r '.id, .name, (.bones | length)'
# Expected:
# bony_pelvis
# Bony Pelvis
# 3

2) 404 path
curl -i http://127.0.0.1:8000/api/boneset/not_a_set
# Expected: HTTP/1.1 404 Not Found
# Body: {"error":"Boneset 'not_a_set' not found"}

3) Sanity check: every bone has image_url and annotations
curl -sS http://127.0.0.1:8000/api/boneset/bony_pelvis \
| jq '.bones[] | select((has("image_url")|not) or (has("annotations")|not))'

# Expected: (no output)

4) Sanity check: every subbone has image_url and annotations
curl -sS http://127.0.0.1:8000/api/boneset/bony_pelvis \
| jq '.bones[] | .subbones[]? | select((has("image_url")|not) or (has("annotations")|not))'

# Expected: (no output)

<img width="676" height="445" alt="Screenshot 2025-09-18 at 12 19 45 AM" src="https://github.com/user-attachments/assets/680456e3-cea8-419f-8217-419c91de39de" />

Taktar commented:
# happy path: returns full JSON
curl -sS http://127.0.0.1:8000/api/boneset/bony_pelvis | jq '.id, .name, (.bones | length)'

# 404 path: non-existent boneset
curl -sS -i http://127.0.0.1:8000/api/boneset/not_a_set | head -n 1
# expect: HTTP/1.1 404 Not Found

# local description fragment
curl -sS 'http://127.0.0.1:8000/api/description?boneId=ischial_ramus'
# expect: <li>…</li> list

Taktar commented:
Manual verification:

1) Start servers: `npm start`
2) Success case:
   curl -sS http://127.0.0.1:8000/api/boneset/bony_pelvis | jq '.id, .name, (.bones | length)'
   Output:
   bony_pelvis
   Bony Pelvis
   3

3) Not-found case:
   curl -sS http://127.0.0.1:8000/api/boneset/not_a_set
   Output:
   {"error":"Boneset 'not_a_set' not found"}

4) Browser check:
   - Visit http://127.0.0.1:8000/api/boneset/bony_pelvis → full JSON returned
   - Visit http://127.0.0.1:8000/api/boneset/not_a_set → 404 JSON error

UcheWendy reviewed (changes requested):
This PR is on the right track, but the missing data in final_boneset.json is a critical issue that needs to be resolved before this can be merged. Please update the JSON to be complete and consider the suggestions for improving the error handling in the server. Great progress so far.

Taktar commented on boneset-api/data/final_bony_pelvis.json:
```diff
@@ -0,0 +1,209 @@
+            "Attachment for the inguinal ligament."
+          ],
+          "image_url": "/images/bony_pelvis/pubic_tubercle.jpg",
+          "annotations": []
```
Are you saying there is missing data in final_boneset.json  or final_bony_pelvis.json? These are two different files right?

Taktar replied:
@UcheWendy  Some of the ischial and pubic sub-bones have no labels in the source slides, so annotations are intentionally []. The merge script will populate these automatically if labels are added later.

UcheWendy replied:
noted

Taktar commented:
Update: boneset data + API wired up

What I changed

Final data file: boneset-api/data/final_bony_pelvis.json now includes:

image_url for bones and sub-bones.

Annotations merged from data_extraction/slide2_pelvis_annotations.json and slide3_annotations.json.

Added/verified labels for: Ilium, Ischium, Pubis and sub-bones Iliac Crest, ASIS, AIIS, PSIS, PIIS, Auricular surface, plus Acetabulum (with position).

Merge helper: added merge_annotations_into_final.py to auto-merge slide annotation JSON into the final file.

Handles id/name matching, normalizes labels, and supports aliases (ASIS, AIIS, etc.).

Re-run anytime after new slides/annotations:

python3 merge_annotations_into_final.py


API:

/api/boneset/:bonesetId serves the local merged file (allowlist + safe path + rate limit).

/api/description?boneId=... now reads descriptions from the merged JSON.

Conflicts resolved: kept the current logic and formatting in templates/js/dropdowns.js and aligned it with description.js (loads description on bone/sub-bone change).

How I verified

JSON is valid
jq . boneset-api/data/final_bony_pelvis.json > /dev/null

Server returns the merged file
npm start
curl -s http://127.0.0.1:8000/api/boneset/bony_pelvis | jq '.id'   # -> "bony_pelvis"

# Spot check a merged sub-bone annotation (example: ASIS)
curl -s http://127.0.0.1:8000/api/boneset/bony_pelvis \
| jq '.bones[] | select(.id=="ilium") | .subbones[] | select(.id=="anterior_superior_iliac_spine") | .annotations[0]'


Remaining gaps / next steps

Some sub-bones under Ischium and Pubis still have empty annotations because the current slides didn’t include those labels. The merge script logs these as “misses.” If we add slides/labels for:

ischial_ramus, ischial_tuberosity, ischial_spine, greater_sciatic_notch, lesser_sciatic_notch,

superior_pubic_ramus, inferior_pubic_ramus, pectineal_line, symphyseal_surface, pubic_tubercle,
I can re-run the merge and they’ll be populated automatically.

Happy to extend the alias map if we see variant label text in additional slides.

Acceptance criteria

✅ final_bony_pelvis.json exists and is complete for the currently available labels and images.

✅ GET http://127.0.0.1:8000/api/boneset/bony_pelvis returns the full JSON.

✅ 404s are returned for unknown bonesets; description endpoint is resilient.

UcheWendy reviewed (approved):
This pull request is now in excellent shape. You've successfully addressed all the major feedback, and the result is a complete data file and a robust, scalable API endpoint. This is a fantastic foundation for the rest of the project. Great work
